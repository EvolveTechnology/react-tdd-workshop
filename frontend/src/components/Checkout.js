import React from "react";
import { Box, Button, Flex, Heading, Image } from "rebass/styled-components";
import { Label, Slider } from "@rebass/forms";
import { CardElement, useStripe, useElements } from "@stripe/react-stripe-js";

import CoffeeCupWithSmile from "assets/coffee_cup_with_smile.png";

export function Checkout({ onSuccess, onError, onCancel }) {
  const [cups, setCups] = React.useState(1);
  const handleSliderChange = e => setCups(e.target.value);

  const stripe = useStripe();
  const elements = useElements();

  const handleSubmit = async event => {
    event.preventDefault();
    const { error, token } = await stripe.createToken(
      elements.getElement(CardElement),
      { currency: "SEK" }
    );

    if (error || !token.id) {
      return onError(error);
    }

    return onSuccess({ token: token?.id, cups });
  };

  const hasCups = cups > 0;
  const enableCheckout = hasCups && stripe;

  return (
    <>
      <Heading>Make a contribution</Heading>
      <Flex flexDirection="column" my={1}>
        <Box>
          <Label htmlFor="coffee-cups" fontSize={3}>
            Quantity ({cups})
          </Label>
          <Slider
            id="coffee-cups"
            name="coffee-cups"
            value={cups}
            min={0}
            max={10}
            onChange={handleSliderChange}
            data-testid="coffee-cups-slider"
          />
          <Flex flexWrap="wrap" marginTop={2} marginBottom={1}>
            {Array.from({ length: cups }, (_, index) => (
              <Image key={index} src={CoffeeCupWithSmile} />
            ))}
          </Flex>
        </Box>
        <form onSubmit={handleSubmit} data-testid="contribute-form">
          <Label htmlFor="percent" my={3} fontSize={3}>
            Pay with Stripe
          </Label>
          <Box maxWidth="400px" margin="0 auto">
            <CardElement
              options={{
                hidePostalCode: true,
                style: {
                  base: {
                    fontFamily: "'Inconsolata', monospace",
                    fontSize: "18px"
                  }
                }
              }}
            />
            <Label my={1} justifyContent="flex-end" color="#ff0000">
              This only a demo. Do not use a real card.
            </Label>
          </Box>
          <Button
            type="submit"
            color="white"
            bg="highlight"
            width="100%"
            fontSize={4}
            marginTop={2}
            disabled={!enableCheckout}
            style={{
              cursor: enableCheckout ? "pointer" : "not-allowed"
            }}
            data-testid="proceed-checkout"
          >
            Checkout
          </Button>
        </form>
        <Button
          onClick={onCancel}
          color="secondary"
          variant="outline"
          fontSize={4}
          marginTop={2}
          style={{ border: "1px solid" }}
          data-testid="cancel-checkout"
        >
          Cancel
        </Button>
      </Flex>
    </>
  );
}
