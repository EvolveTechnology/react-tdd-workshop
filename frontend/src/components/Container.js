import React from "react";

import { Box } from "rebass/styled-components";

export const Container = ({ children }) => (
  <Box
    width={["100%", "90%"]}
    maxWidth="1024px"
    minHeight="100vh"
    margin="0 auto"
    bg="background"
  >
    {children}
  </Box>
);
