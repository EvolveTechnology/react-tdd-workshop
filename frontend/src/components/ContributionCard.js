import React from "react";
import {
  Box,
  Card,
  Flex,
  Heading,
  Image,
  Text
} from "rebass/styled-components";

import CoffeeToGo from "assets/coffee_to_go.png";

export function ContributionCard() {
  return (
    <Card
      bg="muted"
      color="highlight"
      my={2}
      mx={2}
      width="90%"
      maxWidth="575px"
      borderRadius={8}
      sx={{ width: [320, 375, 425, 600], borderRadius: 8 }}
      data-testid="contribution-card"
    >
      <Flex justifyContent="space-around">
        <Flex flex={4} flexDirection="column" p={1}>
          <Flex>
            <Heading fontSize={[3, 4]}>Anonymous</Heading>
          </Flex>
          <Text fontSize={3} as="p" style={{ wordBreak: "break-word" }}>
            Hi there! Keep up the awesome work!
          </Text>
        </Flex>

        <Flex alignItems="center" mx={1}>
          <Box
            fontSize={5}
            sx={{
              display: "flex",
              alignItems: "flex-end",
              color: "white",
              bg: "primary",
              px: 2,
              py: 1,
              borderRadius: 9999
            }}
          >
            <p>+4</p>
            <Image src={CoffeeToGo} width="44px" />
          </Box>
        </Flex>
      </Flex>
    </Card>
  );
}
