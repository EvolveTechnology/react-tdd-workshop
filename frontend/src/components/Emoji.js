import React from "react";
import styled from "styled-components";

const StyledSpan = styled.span`
  vertical-align: middle;
`;

export const Emoji = ({ label, symbol }) => (
  <StyledSpan role="img" aria-label={label}>
    {symbol}
  </StyledSpan>
);
