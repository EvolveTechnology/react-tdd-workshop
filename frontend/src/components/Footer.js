import React from "react";
import { Flex, Image, Text } from "rebass/styled-components";

import CoffeeToGo from "assets/coffee_to_go.png";
import DebitCard from "assets/debit_card.png";
import ReactLogo from "assets/react_logo.png";

export function Footer() {
  return (
    <>
      <Flex py={4} justifyContent="center">
        <Image src={DebitCard} />
        <Image src={CoffeeToGo} />
        <Image src={ReactLogo} width="64px" />
      </Flex>
      <Flex fontSize={2}>
        <Text
          as="a"
          href="https://icons8.com/"
          rel="noopener noreferrer"
          margin="0 auto 1em"
          style={{ textDecoration: "underline" }}
        >
          Icons thanks to Icons8
        </Text>
      </Flex>
    </>
  );
}
