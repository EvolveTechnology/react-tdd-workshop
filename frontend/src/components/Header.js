import React from "react";
import { Text } from "rebass/styled-components";

export function Header() {
  return (
    <>
      <Text m={2} fontSize={3} as="p" textAlign="center">
        Buy me a coffee is a way for you to support my work.
      </Text>
      <Text m={2} fontSize={3} as="p" textAlign="center">
        So far,{" "}
        <Text color="highlight" as="span" fontWeight="bold">
          3
        </Text>{" "}
        people have donated a total of{" "}
        <Text color="highlight" as="span" fontWeight="bold">
          4440
        </Text>{" "}
        cups of coffee.
      </Text>
    </>
  );
}
