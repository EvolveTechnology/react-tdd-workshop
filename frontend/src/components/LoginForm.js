import React from "react";
import { NavLink } from "react-router-dom";
import { Box, Button, Flex, Heading, Text } from "rebass/styled-components";

import { Input, Label } from "@rebass/forms";

export function LoginForm({ onSuccess, onError }) {
  const emailRef = React.useRef();
  const passwordRef = React.useRef();

  const handleSubmit = e => {
    e.preventDefault();

    const email = emailRef.current ? emailRef.current.value : "";
    const password = passwordRef.current ? passwordRef.current.value : "";

    if (email.trim() && password.trim()) {
      return onSuccess({
        email: email.trim(),
        password: password.trim()
      });
    }

    return onError({ message: "Invalid User Input" });
  };

  return (
    <Flex flexDirection="column" my={1} alignItems="center">
      <Heading>Log in</Heading>
      <Box my={2} sx={{ "> form > *": { margin: "0.5em 0" } }}>
        <form data-testid="login-form" onSubmit={handleSubmit}>
          <Label htmlFor="email" fontSize={3}>
            Email
          </Label>
          <Input
            id="email"
            name="email"
            type="email"
            placeholder="jane@example.com"
            fontSize={2}
            maxWidth="300px"
            autoComplete="email"
            ref={emailRef}
            data-testid="login-email"
          />
          <Label htmlFor="password" fontSize={3}>
            Password
          </Label>
          <Input
            id="password"
            name="password"
            type="password"
            fontSize={2}
            maxWidth="300px"
            placeholder="password"
            autoComplete="current-password"
            data-testid="login-password"
            ref={passwordRef}
          />
          <Button
            type="submit"
            width="100%"
            color="black"
            data-testid="submit-login"
          >
            Login
          </Button>
        </form>
      </Box>
      <Text
        my={1}
        fontSize={2}
        textAlign="center"
        style={{ textDecoration: "underline" }}
      >
        <NavLink to="/request" data-testid="request-reset">
          forgot password
        </NavLink>
      </Text>
    </Flex>
  );
}
