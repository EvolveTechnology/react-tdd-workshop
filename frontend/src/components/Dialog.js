import React from "react";
import { Button, Flex, Image } from "rebass/styled-components";

import { Dialog as ReachDialog } from "@reach/dialog";

import { Checkout } from "components/Checkout";
import { LoginForm } from "components/LoginForm";

import CoffeeCup from "assets/coffee_cup.png";

export function Dialog() {
  const [showDialog, setShowDialog] = React.useState(false);
  const open = () => setShowDialog(true);
  const close = () => setShowDialog(false);

  return (
    <>
      <Image
        src={CoffeeCup}
        my={1}
        sx={{
          width: ["40%", "30%", "25%", "20%", "15%"],
          cursor: "pointer"
        }}
        onClick={open}
      />
      <Button
        onClick={open}
        color="white"
        bg="highlight"
        fontSize={5}
        data-testid="donate-button"
      >
        Donate
      </Button>
      <ReachDialog
        aria-label="Donate"
        isOpen={showDialog}
        onDismiss={close}
        data-testid="donate-dialog"
      >
        <Flex flexDirection="column">
          <LoginForm onSuccess={console.log} onError={console.log} />
          <Checkout
            onSuccess={console.log}
            onError={console.log}
            onCancel={close}
          />
        </Flex>
      </ReachDialog>
    </>
  );
}
