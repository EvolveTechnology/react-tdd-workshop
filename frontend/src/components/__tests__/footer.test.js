import React from "react";

import { Footer } from "components/Footer";
import { mount } from "enzyme";

import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

describe("Credit icon providers", () => {
  const { getByText } = render(<Footer />);
  const element = getByText(/icons8/i);

  it("Mentions the icon providers", () => {
    expect(element).toBeInTheDocument();
  });

  it("Links to the icon providers", () => {
    expect(element.getAttribute("href")).toMatch("icons8.com");
  });
});

test("Credit style", () => {
  let app = mount(<Footer />);
  expect(app.html()).toMatchSnapshot();
});
