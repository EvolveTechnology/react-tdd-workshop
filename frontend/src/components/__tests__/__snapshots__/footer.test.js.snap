// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Credit style 1`] = `"<div class=\\"sc-AxjAm sc-AxirZ coteZM\\"><img src=\\"debit_card.png\\" class=\\"sc-AxjAm fFIAuE\\"><img src=\\"coffee_to_go.png\\" class=\\"sc-AxjAm fFIAuE\\"><img src=\\"react_logo.png\\" width=\\"64px\\" class=\\"sc-AxjAm cfWTjF\\"></div><div font-size=\\"2\\" class=\\"sc-AxjAm sc-AxirZ bsLIVh\\"><a href=\\"https://icons8.com/\\" rel=\\"noopener noreferrer\\" style=\\"text-decoration: underline;\\" class=\\"sc-AxjAm gWKrNW\\">Icons thanks to Icons8</a></div>"`;
